### Задание 1

1. Как критики относятся к спортивным играм?
2. Критикам нравятся больше игры на PC или на PS4?
3. Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

* сформулируйте нулевую и альтернативную гипотезы
* выберите пороговый уровень статистической значимости
* опишите полученные результаты статистического теста

In [3]:
import pandas as pd
from scipy import stats as st

In [3]:
df = pd.read_csv('video_games_sales.csv')
df.head()

Rank                           Name                      basename  \
0     1                     Wii Sports                    wii-sports   
1     2              Super Mario Bros.              super-mario-bros   
2     3                 Mario Kart Wii                mario-kart-wii   
3     4  PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4     5              Wii Sports Resort             wii-sports-resort   

      Genre ESRB_Rating Platform         Publisher         Developer  \
0    Sports           E      Wii          Nintendo      Nintendo EAD   
1  Platform         NaN      NES          Nintendo      Nintendo EAD   
2    Racing           E      Wii          Nintendo      Nintendo EAD   
3   Shooter         NaN       PC  PUBG Corporation  PUBG Corporation   
4    Sports           E      Wii          Nintendo      Nintendo EAD   

   VGChartz_Score  Critic_Score  ...  NA_Sales  PAL_Sales  JP_Sales  \
0             NaN           7.7  ...       NaN        NaN       NaN   
1             NaN          10.0  ...       NaN        NaN       NaN   
2             NaN           8.2  ...       NaN        NaN       NaN   
3             NaN           NaN  ...       NaN        NaN       NaN   
4             NaN           8.0  ...       NaN        NaN       NaN   

   Other_Sales    Year  Last_Update  \
0          NaN  2006.0          NaN   
1          NaN  1985.0          NaN   
2          NaN  2008.0  11th Apr 18   
3          NaN  2017.0  13th Nov 18   
4          NaN  2009.0          NaN   

                                                 url  status Vgchartzscore  \
0  http://www.vgchartz.com/game/2667/wii-sports/?...       1           NaN   
1  http://www.vgchartz.com/game/6455/super-mario-...       1           NaN   
2  http://www.vgchartz.com/game/6968/mario-kart-w...       1           8.7   
3  http://www.vgchartz.com/game/215988/playerunkn...       1           NaN   
4  http://www.vgchartz.com/game/24656/wii-sports-...       1           8.8   

                                         img_url  
0  /games/boxart/full_2258645AmericaFrontccc.jpg  
1                   /games/boxart/8972270ccc.jpg  
2  /games/boxart/full_8932480AmericaFrontccc.jpg  
3  /games/boxart/full_8052843AmericaFrontccc.jpg  
4  /games/boxart/full_7295041AmericaFrontccc.jpg  

[5 rows x 23 columns]

In [ ]:
# Как критики относятся к спортивным играм?
# H0 - оценки критиков аналогичны другим категориям игр, H1 - средние оценки критиков по спортивным играм выше, чем в других жанрах

In [22]:
df_clean = df[df.Critic_Score.notnull()]
df_clean.head(50)

Rank                                          Name  \
0      1                                    Wii Sports   
1      2                             Super Mario Bros.   
2      3                                Mario Kart Wii   
4      5                             Wii Sports Resort   
5      6            Pokemon Red / Green / Blue Version   
6      7                         New Super Mario Bros.   
8      9                     New Super Mario Bros. Wii   
9     10                                     Minecraft   
11    12                                      Wii Play   
12    13                            Kinect Adventures!   
13    14                                    Nintendogs   
14    15                                 Mario Kart DS   
15    16                 Pokemon Gold / Silver Version   
16    17                                       Wii Fit   
17    18                                  Wii Fit Plus   
18    19                             Super Mario World   
19    20                            Grand Theft Auto V   
20    21                            Grand Theft Auto V   
21    22  Brain Age: Train Your Brain in Minutes a Day   
24    25                                  Mario Kart 7   
25    26               Pokemon Diamond / Pearl Version   
26    27                 Grand Theft Auto: San Andreas   
28    29                                   Pokemon X/Y   
29    30               Pokemon Ruby / Sapphire Version   
30    31                   Grand Theft Auto: Vice City   
31    32                              Pokemon Sun/Moon   
33    34                 Pokemon Black / White Version   
35    36                           Mario Kart 8 Deluxe   
36    37                        Counter-Strike: Source   
38    39                        Gran Turismo 3: A-Spec   
39    40   Brain Age 2: More Training in Minutes a Day   
40    41                Call of Duty: Modern Warfare 3   
41    42                       Call of Duty: Black Ops   
42    43       Pokemon Yellow: Special Pikachu Edition   
43    44                                        Halo 3   
45    46                         Red Dead Redemption 2   
48    49                           Super Mario Odyssey   
49    50                Call of Duty: Modern Warfare 2   
50    51                            Call of Duty: WWII   
51    52                Call of Duty: Modern Warfare 3   
52    53                       Super Smash Bros. Brawl   
53    54                          Grand Theft Auto III   
54    55                       New Super Mario Bros. 2   
55    56                                      Portal 2   
57    58                            Super Mario Galaxy   
58    59      Pokemon Heart Gold / Soul Silver Version   
59    60                       Call of Duty: Black Ops   
61    62                           Super Mario 3D Land   
62    63                     Animal Crossing: New Leaf   
63    64                    Super Smash Bros. Ultimate   

                                       basename             Genre ESRB_Rating  \
0                                    wii-sports            Sports           E   
1                              super-mario-bros          Platform         NaN   
2                                mario-kart-wii            Racing           E   
4                             wii-sports-resort            Sports           E   
5                                    pokmon-red      Role-Playing           E   
6                          new-super-mario-bros          Platform           E   
8                      new-super-mario-bros-wii          Platform           E   
9                                     minecraft              Misc         NaN   
11                                     wii-play              Misc           E   
12                            kinect-adventures             Party           E   
13                                   nintendogs        Simulation           E   
14                                mario-kart-ds            Racing        

In [51]:
genre_list = list(df.Genre.unique())
genre_list

['Sports',
 'Platform',
 'Racing',
 'Shooter',
 'Role-Playing',
 'Puzzle',
 'Misc',
 'Party',
 'Simulation',
 'Action',
 'Action-Adventure',
 'Fighting',
 'Strategy',
 'Adventure',
 'Music',
 'MMO',
 'Sandbox',
 'Visual Novel',
 'Board Game',
 'Education']

In [50]:
def comparison_genres(genre):
    
    result = st.ttest_ind(df_clean[df_clean.Genre == 'Sports'].Critic_Score, df_clean[df_clean.Genre == genre].Critic_Score, equal_var=False)

    return result

In [49]:
dict_score = {}
for genre in genre_list:
    
    if comparison_genres(genre).pvalue < alpha:
        
        dict_score[genre] = comparison_genres(genre)

data_ab = pd.DataFrame(dict_score)
data_ab.index = ['statistic', 'pvalue']

print('Отвергаем нулевую гипотезу, критики по разному относятся к этим жанрам в сравнении со спортивными играми')
data_ab

Отвергаем нулевую гипотезу, критики по разному относятся к этим жанрам в сравнении со спортивными играми


/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3584: RuntimeWarning: Degrees of freedom <= 0 for slice
  **kwargs)
/opt/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py:209: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)


Racing  Role-Playing      Misc     Party  Simulation  \
statistic  2.223218     -2.436431  4.064501  4.771709    3.138295   
pvalue     0.026431      0.014964  0.000054  0.000230    0.001834   

                 Action  Action-Adventure     Music  
statistic  6.046767e+00         -2.928071 -3.200873  
pvalue     1.851213e-09          0.003981  0.003347

#### Выводы:

1. самое большое различие в оценках критиков по спортивным и экшен играм
2. оценки спортивных игр выше только по сравнению: Role-Playing, Action-Adventure, Music
3. не можем отвергнуть нулевую гипотезу: 'Platform', 'Shooter', 'Puzzle', 'Fighting', 'Strategy', 'Adventure', 'Music', 'MMO', 'Sandbox', 'Visual Novel', 'Board Game', 'Education'

In [53]:
# Критикам нравятся больше игры на PC или на PS4?
# H0 - оценки критиков аналогичны для этих платформ, H1 - средние оценки критиков PC выше, чем на PS4

In [54]:
pc = df_clean[df_clean.Platform == 'PC']
ps4 = df_clean[df_clean.Platform == 'PS4']

In [62]:
result = st.ttest_ind(pc.Critic_Score, ps4.Critic_Score, equal_var=False)

if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, критики по разному оценивают игры на PC или на PS4')
else:
    print('Не отвергаем нулевую гипотезу')
result

Отвергаем нулевую гипотезу, критики по разному оценивают игры на PC или на PS4


Ttest_indResult(statistic=-2.7394476056351627, pvalue=0.006931808250254211)

#### Выводы:

1. Критики по разному оценивают игры на PC или на PS4
2. Критикам немного больше нравятся игры на PS4

In [68]:
# Критикам больше нравятся стрелялки или стратегии?
# H0 - оценки критиков аналогичны для этих жанров, H1 - средние оценки критиков игр в жанре 'Shooter' выше, чем 'Strategy'

In [66]:
shoot = df_clean[df_clean.Genre == 'Shooter']
strat = df_clean[df_clean.Genre == 'Strategy']

In [69]:
result = st.ttest_ind(shoot.Critic_Score, strat.Critic_Score, equal_var=False)

if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, критики по разному оценивают игры на PC или на PS4')
else:
    print('Не отвергаем нулевую гипотезу')
result

Не отвергаем нулевую гипотезу


Ttest_indResult(statistic=-1.6073949711166526, pvalue=0.10838786414223071)

#### Выводы:

pvalue в 2 раза выше alfa - нулевую гипотезу о схожести оценок этих жанров отвергнуть не можем

In [12]:
import glob
import re
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from gensim import corpora
from gensim.models import TfidfModel
from gensim import similarities
import seaborn as sns
from scipy.cluster import hierarchy

In [161]:
spam_files = pd.read_csv('spam.csv')
spam_files_copy = spam_files.copy()
spam_files_copy.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [162]:
# приведем все слова к нижнему регистру
spam_files['Message'] = spam_files.Message.str.lower()
spam_files.head()

Category                                            Message
0      ham  go until jurong point, crazy.. available only ...
1      ham                      ok lar... joking wif u oni...
2     spam  free entry in 2 a wkly comp to win fa cup fina...
3      ham  u dun say so early hor... u c already then say...
4      ham  nah i don't think he goes to usf, he lives aro...

In [163]:
# удалим мусорные символы
spam_files['Message'] = spam_files.Message.replace('[\W_]+',' ', regex=True)
spam_files.head()

Category                                            Message
0      ham  go until jurong point crazy available only in ...
1      ham                           ok lar joking wif u oni 
2     spam  free entry in 2 a wkly comp to win fa cup fina...
3      ham       u dun say so early hor u c already then say 
4      ham  nah i don t think he goes to usf he lives arou...

In [164]:
# удалим стоп-слова
stop = set(stopwords.words('english'))
spam_files['Message'] = spam_files.Message.apply(lambda x: [word for word in x.split() if word not in stop])
spam_files.head()

Category                                            Message
0      ham  [go, jurong, point, crazy, available, bugis, n...
1      ham                     [ok, lar, joking, wif, u, oni]
2     spam  [free, entry, 2, wkly, comp, win, fa, cup, fin...
3      ham      [u, dun, say, early, hor, u, c, already, say]
4      ham     [nah, think, goes, usf, lives, around, though]

In [165]:
# приведем все слова к нормальной форме
wordnet_lemmatizer = WordNetLemmatizer()
spam_files['Message'] = spam_files.Message.apply(lambda x: ' '.join([wordnet_lemmatizer.lemmatize(word) for word in x]))

In [166]:
spam_files.head()

Category                                            Message
0      ham  go jurong point crazy available bugis n great ...
1      ham                            ok lar joking wif u oni
2     spam  free entry 2 wkly comp win fa cup final tkts 2...
3      ham                u dun say early hor u c already say
4      ham                nah think go usf life around though

In [167]:
# преобразуем все сообщения в вектора TF-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(spam_files.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)

In [168]:
tfidf_matrix['Category'] = spam_files.Category
tfidf_matrix.head()

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

   0125698789   02  ...  zindgi  zoe  zogtorius  zoom  zouk  zyada   èn   ú1  \
0         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0   
1         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0   
2         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0   
3         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0   
4         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0  0.0   

   〨ud  Category  
0  0.0       ham  
1  0.0       ham  
2  0.0      spam  
3  0.0       ham  
4  0.0       ham  

[5 rows x 8049 columns]

In [169]:
# разделим данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. 
from sklearn.model_selection import train_test_split
data_train, data_test = train_test_split(tfidf_matrix, test_size=0.3, random_state=42)

In [170]:
data_train

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
708   0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4338  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5029  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4921  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2592  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
3772  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5191  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5226  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5390  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
860   0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zindgi  zoe  zogtorius  zoom  zouk  zyada   èn  \
708          0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
4338         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
5029         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
4921         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
2592         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
...          ...  ...  ...     ...  ...        ...   ...   ...    ...  ...   
3772         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
5191         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
5226         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
5390         0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   
860          0.0  0.0  ...     0.0  0.0        0.0   0.0   0.0    0.0  0.0   

       ú1  〨ud  Category  
708   0.0  0.0       ham  
4338  0.0  0.0       ham  
5029  0.0  0.0       ham  
4921  0.0  0.0       ham  
2592  0.0  0.0       ham  
...   ...  ...       ...  
3772  0.0  0.0       ham  
5191  0.0  0.0       ham  
5226  0.0  0.0       ham  
5390  0.0  0.0       ham  
860   0.0  0.0       ham  

[3900 rows x 8049 columns]

In [171]:
# удалим из тестовой и тренировочной выборок категорию
y_train, y_test = data_train['Category'], data_test['Category']
del data_train['Category']
del data_test['Category']

In [172]:
from sklearn.linear_model import LogisticRegression
LR = LogisticRegression()

In [173]:
# построим модель логистической регрессии, при random_state=42
clf = LR.fit(data_train, y_train)
clf

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [175]:
# оценим точность модели на тестовых данных
print(clf.score(data_test, y_test))

divination = clf.predict(data_test)
result = pd.DataFrame([y_test.values, divination]).T
result.head()

0.958732057416268


0    1
0  ham  ham
1  ham  ham
2  ham  ham
3  ham  ham
4  ham  ham

In [176]:
# опишем результаты при помощи confusion_matrix
from sklearn.metrics import confusion_matrix

tn, fp, fn, tp = confusion_matrix(y_test.values, divination).ravel()
tn, fp, fn, tp

(1445, 3, 66, 158)

#### Результат 
* 158 истинных положительных результатов, 
* 3 ложных положительных результатов, 
* 66 ложных отрицательных и 
* 1445 истинных отрицательных

In [187]:
# Постройте датафрейм, который будет содержать все исходные тексты сообщений, 
# классифицированные неправильно (с указанием фактического и предсказанного)
result['Message'] = spam_files_copy.Message
error = result.loc[(result[0] != result[1])]
error.rename(columns={0: 'fact', 1: 'predict'}, inplace=True)
error

fact predict                                            Message
17     ham    spam  Eh u remember how 2 spell his name... Yes i di...
40    spam     ham  Pls go ahead with watts. I just wanted to be s...
47    spam     ham                    Fair enough, anything going on?
74    spam     ham                               U can call me now...
84    spam     ham                                     Yup next stop.
...    ...     ...                                                ...
1525  spam     ham                      Hi:)cts employee how are you?
1567  spam     ham                      I am late. I will be there at
1569  spam     ham  Today is ACCEPT DAY..U Accept me as? Brother S...
1576   ham    spam  Sounds gd... Haha... Can... Wah, u yan jiu so ...
1586  spam     ham  I was wondering if it would be okay for you to...

[69 rows x 3 columns]